### PR TITLE
Replace webhooks with CEL validation rules for LocalQueue

### LINKED ISSUE #463: remove the webhooks in favor of CEL once 1.24 reaches EoL
**What would you like to be cleaned**:
we can remove the webhooks in favor of CEL once 1.24 reaches EoL https://kubernetes.io/releases/

**Why is this needed**:
Since Kubernetes 1.25, we can use CEL validation rules to implement a few common immutability patterns directly in the manifest for a CRD.
ref: https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/

### PR DESCRIPTION
**Pull request by IrvingMg**
#### What type of PR is this?

/kind cleanup
#### What this PR does / why we need it:

As mentioned in #463, we can enforce immutability and other validations using CEL rules instead of webhooks. Thus, this PR removes the validating webhooks for the `localqueue` type and adds the corresponding validation rules.

#### Which issue(s) this PR fixes:

Relates to #463 

#### Does this PR introduce a user-facing change?

```release-note
Replaced LocalQueue admission webhook with CRD validation rules.

ACTION REQUIRED: Requires Kubernetes 1.25 or newer
```

### CODE REVIEW

File: apis/kueue/v1beta1/localqueue_types.go [hunk 1]
```diff
@@ -25,10 +25,13 @@ import (
 // LocalQueueSpec defines the desired state of LocalQueue
 type LocalQueueSpec struct {
 	// clusterQueue is a reference to a clusterQueue that backs this localQueue.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="field is immutable"
 	ClusterQueue ClusterQueueReference `json:"clusterQueue,omitempty"`
 }
 
 // ClusterQueueReference is the name of the ClusterQueue.
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
```
Comment (alculquicondor) [inline apis/kueue/v1beta1/localqueue_types.go:34]: @jpbetz there is no CEL library for common k8s regexes, is there?

File: pkg/util/testing/not_found_error_cmp.go
Comment (alculquicondor) [inline pkg/util/testing/not_found_error_cmp.go:1]: It looks like we should change the name of the file. Maybe `error_matchers.go`

File: pkg/util/testing/not_found_error_cmp.go [hunk 2]
```diff
@@ -75,3 +75,27 @@ func (matcher *isForbiddenErrorMatch) FailureMessage(actual interface{}) (messag
 func (matcher *isForbiddenErrorMatch) NegatedFailureMessage(actual interface{}) (message string) {
 	return format.Message(actual, "not to be Forbidden error")
 }
+
+func BeInvalidError() types.GomegaMatcher {
+	return &isInvalidErrorMatch{}
```
Comment (alculquicondor) [inline pkg/util/testing/not_found_error_cmp.go:80]: Maybe we can have a single class that implements all the matchers by accepting one of the apierrors functions and a name.

**Comment by alculquicondor** [comment]
By the way, this requires a release note.

Also, please include the text `ACTION REQUIRED`, followed by a request to use kubernetes 1.25 or newer https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/

File: pkg/util/testing/error_matchers.go [hunk 3]
```diff
@@ -76,26 +76,47 @@ func (matcher *isForbiddenErrorMatch) NegatedFailureMessage(actual interface{})
...
+)
+
+func (em errorMatcher) String() string {
+	return []string{"NotFoundError", "ForbiddenError", "InvalidError"}[em]
 }
 
-type isInvalidErrorMatch struct {
+type apiError func(error) bool
+
+func (em errorMatcher) IsApiError(err error) bool {
```
Comment (alculquicondor) [inline pkg/util/testing/error_matchers.go:93]:
```suggestion
func (em errorMatcher) isApiError(err error) bool {
```

File: pkg/util/testing/error_matchers.go [hunk 4]
```diff
@@ -76,26 +76,47 @@ func (matcher *isForbiddenErrorMatch) NegatedFailureMessage(actual interface{})
...
+	return []apiError{apierrors.IsNotFound, apierrors.IsForbidden, apierrors.IsInvalid}[em](err)
+}
+
+type isErrorMatch struct {
+	name errorMatcher
+}
+
+func BeError(name errorMatcher) types.GomegaMatcher {
```
Comment (alculquicondor) [inline pkg/util/testing/error_matchers.go:101]:
```suggestion
func BeAPIError(name errorMatcher) types.GomegaMatcher {
```

File: pkg/util/testing/error_matchers.go [hunk 5]
```diff
@@ -76,26 +76,47 @@ func (matcher *isForbiddenErrorMatch) NegatedFailureMessage(actual interface{})
 	return format.Message(actual, "not to be Forbidden error")
 }
 
-func BeInvalidError() types.GomegaMatcher {
-	return &isInvalidErrorMatch{}
+type errorMatcher int
```
Comment (alculquicondor) [inline pkg/util/testing/error_matchers.go:79]: You could also reimplement `BeForbiddenError` as `return BeAPIError(ForbiddenError)`. Similarly for `BeNotFoundError`

File: apis/kueue/v1beta1/localqueue_types.go (hunk 1 above)
Comment (jpbetz) [reply apis/kueue/v1beta1/localqueue_types.go:34]: Not yet, but @alexzielenski is working on adding them as [part of declarative validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4153-declarative-validation#formats), both OpenAPI formats and CEL utility functions to check common regexes (like dns1123label) are planned.

**Comment by alculquicondor** [comment]
/release-note-edit

```release-note
Replaced LocalQueue admission webhook with CRD validation rules.

ACTION REQUIRED: Requires Kubernetes 1.25 or newer
```

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Added CRD validation rules to LocalQueue.

ACTION REQUIRED: Requires Kubernetes 1.25 or newer
```

Just to match the format of other release notes.

**Pull request merged**